packages/vite-plugin-cloudflare/src/plugin-config.ts

```diff
@@ -1,101 +1,189 @@
+import * as fs from 'node:fs';
 import * as path from 'node:path';
 import * as vite from 'vite';
-import { unstable_getMiniflareWorkerOptions } from 'wrangler';
-import type { AssetConfig } from './assets';
-import type { SourcelessWorkerOptions } from 'wrangler';
+import { readConfig } from 'wrangler';
+import { invariant } from './shared';
+import type { Config } from 'wrangler';
 
-export interface WorkerOptions {
-	main: string;
+export interface PluginConfig {
 	wranglerConfig?: string;
-	// TODO: tighten up types so assets can only be bound to entry worker
-	assetsBinding?: string;
-	overrides?: vite.EnvironmentOptions;
+	viteEnvironmentName?: string;
+	auxiliaryWorkers?: Array<{
+		wranglerConfig: string;
+		viteEnvironmentName?: string;
+	}>;
 }
 
-export interface PluginConfig<
-	TWorkers extends Record<string, WorkerOptions> = Record<
-		string,
-		WorkerOptions
-	>,
-	TEntryWorker extends string = Extract<keyof TWorkers, string>,
-> {
-	workers?: TWorkers;
-	entryWorker?: TEntryWorker;
-	assets?: AssetConfig;
-	persistTo?: string | false;
+type Defined<T> = Exclude<T, undefined>;
+
+export type AssetsOnlyConfig = Config & { assets: Defined<Config['assets']> };
+
+export type WorkerConfig = Config & {
+	name: Defined<Config['name']>;
+	main: Defined<Config['main']>;
+};
+
+interface AssetsOnlyResult {
+	type: 'assets-only';
+	config: AssetsOnlyConfig;
 }
 
-export interface NormalizedPluginConfig {
-	workers: Record<
-		string,
-		{
-			entryPath: string;
-			wranglerConfigPath: string;
-			assetsBinding?: string;
-			workerOptions: SourcelessWorkerOptions & { name: string };
-		}
-	>;
-	entryWorkerName?: string;
-	assets: AssetConfig;
-	persistPath: string | false;
+interface WorkerResult {
+	type: 'worker';
+	config: WorkerConfig;
+}
+
+interface BasePluginConfig {
 	wranglerConfigPaths: Set<string>;
 }
 
-const DEFAULT_PERSIST_PATH = '.wrangler/state/v3';
+interface AssetsOnlyPluginConfig extends BasePluginConfig {
+	type: 'assets-only';
+	config: AssetsOnlyConfig;
+}
+
+interface WorkersPluginConfig extends BasePluginConfig {
+	type: 'workers';
+	workers: Record<string, WorkerConfig>;
+	entryWorkerEnvironmentName: string;
+	wranglerConfigPaths: Set<string>;
+}
+
+export type ResolvedPluginConfig = AssetsOnlyPluginConfig | WorkersPluginConfig;
+
+function getConfigResult(
+	configPath: string,
+	userConfig: vite.UserConfig,
+	wranglerConfigPaths: Set<string>,
+	isEntryWorker?: boolean,
+): AssetsOnlyResult | WorkerResult {
+	if (wranglerConfigPaths.has(configPath)) {
+		throw new Error(`Duplicate Wrangler config path found: ${configPath}`);
+	}
+
+	const wranglerConfig = readConfig(configPath, {
+		// We use the mode from the user config rather than the resolved config for now so that the mode has to be set explicitly. Otherwise, some things don't work as expected when `development` and `production` environments are not present.
```

Comment:
why is this? 😕

Reply:
It's the way `readConfig` from Wrangler works, unfortunately. I was hoping that if an environment is specified and no environments are present in the config file then it would just use the top level config. This doesn't seem to be the case though.

Reply:
> I was hoping that if an environment is specified and no environments are present in the config file then it would just use the top level config

I am pretty sure that is how it works. It just warns but doesn't error: https://github.com/cloudflare/workers-sdk/blob/b5c72af7ea061cae9948ae51e3252d8857643890/packages/wrangler/src/config/validation.ts#L245

Reply:
If that was the case then you'd be able to the change `env` here from `userConfig.mode` to some arbitrary value and all the tests would still pass.

Reply:
This comment explains the behaviour here:

https://github.com/cloudflare/workers-sdk/blob/bea65583f72fcf8346ed82d4b9b1176669bcfd23/packages/wrangler/src/config/validation.ts#L193-L205

For Workers, if an environment name is specified and the environment is not configured then a fake empty environment is used in its place. This has the effect of removing inheritable properties that are set at the top level. You can see the relevant code here (the third argument is the dummy env):

https://github.com/cloudflare/workers-sdk/blob/bea65583f72fcf8346ed82d4b9b1176669bcfd23/packages/wrangler/src/config/validation.ts#L219-L228

What we really want is the behaviour that is used for Pages but it feels like this might be a breaking change.